### DISCUSSION TITLE
Stateful connection configuration

### DISCUSSION THREAD

**Discussion by tailhook** [geldata/gel#1896] [Pre-RFC]
# Abstract

This proposal describes how stateful changes of the connection are implemented in a way that allows seamlessly re-establish underlying connection.

# API

## Connection Object and Pool

Instead of exposing a TCP connection to the user we use a set wrapper classes:

```python
class Connection:
    _impl: _SingletonPool
    _desired_config: SessionConfig

class _SingletonPool
    _impl: RawConnection

class RawConnection:
    _protocol: BlockingIOProtocol
    _current_config: SessionConfig

def connect() -> Connection: ...
```

Similarly we update connection pool classes API:

```python
class Pool:
    _impl: _PoolImpl
    _desired_config: SessionConfig

class _PoolImpl:
    _connections: deque[RawConnection]  # Simplified
```

`RawConnection` class is exposed for users want to do their own session config management stuff, connection pooling or control of reconnection.

Both ``Pool`` and ``Connection`` have:

1. `execute`, `query`, `query_one` set of query functions (implement `Executor` abstract class from RFC1004). All the methods reconnect if called on broken connection. Read-only queries broken in the middle of the query are retried.
2. `retry` method (as described in RFC1004)
3. `try_transaction` method for more fine-grained control over transactions

## Modification methods

The following things are modified by a method that returns a distinct object of the same type and same underlying network connections but distinct config (i.e. `Connection` retains the same `_SingletonPool` and `Pool` retains the same `_PoolImpl`):

1. Asking for a read-only access
2. Setting session configuration
3. Transaction options
4. Setting tracing/debugging metadata for queries (this is a part of the spec just to show the API in perspective, metadata will be implemented later).

Instead of applying settings immediately, the parameters used are only applied when query methods are actually executed. Some optimizations are done:

1. If connection already has the config specified it's reused without modifying session config.
2. Connection pool keeps track of what session changes are done and picks connections that has these settings applied, if possible.

Here are method signatures:

```python
class Connection:
    ...
    def read_only(self, primary: bool = false) -> Connection: ...
    def with_mutations(self, primary: bool = false) -> Connection: ...
    def with_session_config(self, **config) -> Connection: ...
    def with_transaction_options(self, isolation: ...) -> Connection: ...
    def with_retry_options(self, attempts: int = 3, ...) -> Connection: ...
    def with_metadata(self, **data) -> Connection: ...

class Pool:
    ...
    def read_only(self, primary: bool = false) -> Pool: ...
    def with_mutations(self, primary: bool = false) -> Pool: ...
    def with_session_config(self, **config) -> Pool: ...
    def with_transaction_options(self, isolation: ...) -> Connection: ...
    def with_retry_options(self, attempts: int = 3, ...) -> Connection: ...
    def with_metadata(self, **data) -> Pool: ...
```

After modification, connection/pool object can be used interchangeably:

```python
conn = edgedb.connect()
read_only_conn = conn.read_only()
conn.execute("INSERT User { ... }", ...)
print(read_only_conn.query("SELECT User"))
read_only_conn.execute("INSERT User { .. }", ...)  # throws an error
read_only_conn.with_mutations().execute("INSERT User { ... }", ...)  # good
```

Or session config example:

```python
conn = edgedb.connect()
conn2 = conn.with_session_config(param="value")
conn.execute("INSERT User { ... }", ...)  # without config
print(conn2.query("SELECT User"))  # with config
conn.execute("INSERT User { .. }", ...)  # without config again
```

# EdgeDB Changes

1. Add a header for `Prepare` to disable `CONFIGURE SESSION` statements, this disables manual modifications of the session state (it's allowed only when using query methods on `RawConnection` directly)
2. Add a header for `Prepare` to only accept read-only queries (already
declared in failover spec, see Notes)

# Alternatives

## Names

`read_only` could be `with_read_only` to support convention. But it looks like it's clear enough.

`with_mutations` could be `read_write` or `with_read_write` or `clear_read_only`. Or it could be `read_only(false)`.

## Is `with_mutations` Needed?

It's intuitive that databases are mutable by default. But there is a large class of applications that are mostly read-only and must have limited and easy to find places having mutations. For those apps it's better to have read-only connection by default and use a pattern like this for writes:

```python
    def save(conn):
        async for tx in conn.with_mutations().retry():
          async with tx:
            ...
```

While `retry` is also easy to find, nothing stops user from writing:

```python
    conn.fetch("INSERT User { .. }")
```

Except the `read_only` configuration.

The metadata is will probably be implemented later and works the same:

```python
def http_handler(req, conn):
    req['conn'] = conn.with_metadata(url=req.url)
    process_request(req)
```

## Use Context Manager

While it's tempting to use context manager instead:

```python

def handler(conn):
    with conn.session_config(name="value"):
        conn.query("...")
```

But behavior implies that session config is reset on each `with/__exit__`, which consequences:

1. When `handler()` is interrupted, e.g. a timeout occurred, connection could be left in the inconsistent state (i.e. should be dropped)
2. It prevents caching the session config, e.g. in the case above: `handler(c); handler(c)` will not set session config second time, but use the knowledge that these settings are already active.

Also it improves clarity and composition. E.g.:

```python
def money_transfer(tx):
    user1 = tx.with_metadata(user='user1')
    user2 = tx.with_metadata(user='user2')
    withdraw(user1)
    deposit(user2)
    log_withdraw(user1)
    log_deposit(user2)
```

(This implies `Transaction` also contains `with_metadata` method, but not others of this spec).

Additional thing considered is that with context manager, if connection is failed in the middle of the block:

```python
    with conn.session_config(name="value"):
        conn.query("...")
        # << -- here
        conn.query("...")
```

The implementation should reconnect and reapply the session configuration anyway. So it deviates from the usual behavior of context manager anyway.

# Notes

This spec intersects with [failover spec](https://github.com/edgedb/edgedb/discussions/1859) and transactions (RFC1004), so maybe merged into [RFC1004](https://github.com/edgedb/rfcs/blob/master/text/1004-transactions-api.rst).

This spec also upgrades the way default retry options are configured on the connection pool object.

Also with this spec we could remove `non_transactional_writes` hint described in failover spec and recommend `pool.read_only()` and `pool.with_mutations().retry()` instead.

**Comment by 1st1** [comment]
> Add a header for Prepare to only accept read-only queries (already
declared in failover spec, see Notes)

Actually I think we should approach this similarly to expected cardinality. The client sends an "expected operations" flagged enum (can group "MUTATION", "CONFIG", etc) and the server errors out if the query is outside of the expected operations.

**Comment by 1st1** [comment]
Few notes:

First, we need to define some soft of "ownership" concept for non-Rust languages. The current proposal outlines method signatures like `with_session_config(self, **config) -> Connection` but doesn't specify *how* the old `Connection` and the new `Connection` are related to each other.  E.g. for the following code in Python:

```python
con1 = connect()
con2 = con1.with_session_config(...)
```

I'd expect that `con1` isn't usable until `con2` is explicitly "unwrapped" or "cancelled". Using them simultaneously with one underlying physical connection would be error-prone and more over not possible with the current protocol. So any operation on `con1`, while `con2` is still active, must be rejected.  Something like:

```python
con1 = connect()
con2 = con1.with_session_config(...)

# Any operations on con1 are prohibited at this point.

print(con2.query(...))

con2.unwrap()  # we might need to find a better name for this method

# Any operations on *con2* are prohibited at this point.

# But working with `con1` is now allowed.
print(con1.query(...))
```


Speaking of physical connections, essentially, the `def connect() -> Connection` API returns a "virtual" connection (essentially a Pool of size 1).  There are a variety of `with_*` methods to "tune" that virtual `Connection` object.  But we must guarantee that for all "virtual" connections created out of one `connect()` call we always have only one physical connection *at a time*.

Second, regarding "Is with_mutations Needed?", I think that the answer is "no". By default connections should be read/write (unless the specific `connect_readonly()`-like) API is used to open them. Doing it the other way - upgrading connection from "readonly" to "read/write" isn't always possible without closing the underlying physical connection and opening a new one. Which can be quite expensive. So I propose to:

* Add an API to open readonly connections when the application knows it may connect to a replica: `connect_readonly -> ReadonlyConnection`. There will be no way to make a readonly connection a read/write one.

* `connect()` by default would return a read/write connection. A read/write connection would have an option to create a readonly variant of it (enforced at the protocol level).

**Reply by tailhook** [reply]
> I'd expect that con1 isn't usable until con2 is explicitly "unwrapped" or "cancelled". Using them simultaneously with one underlying physical connection would be error-prone and more over not possible with the current protocol.

Well, unwrapping would clearly be error prone. There is `with` statement for that in python, but this is different from what this RFC proposes.

The idea is that it's used like this:
```
def hander(req, base_conn, *args):
    user_conn = base_conn.with_session_config(username=req.session['username'])
    handle_user_request(req, user_conn, *args)
```
So no risk of using wrong connection in `handle_user_request`

>  But we must guarantee that for all "virtual" connections created out of one connect() call we always have only one physical connection at a time.

Why? The idea is that all of them can. If some can't we do that as a parameter to `connect()`. The "readonly" and "metadata" things are actually headers for `Execute` message, so set on per packet basis. So we could put it into `query()` signature if we want, it's just too weird to overload signature of that function too much.

Session data is accumulated, and applied before applying request. Then we can use different approaches to cache connections having that specific session configuration.

> Doing it the other way - upgrading connection from "readonly" to "read/write" isn't always possible without closing the underlying physical connection and opening a new one.

Yes, but I'm not saying that it's always possible. If you connect to read-only edgedb instance, upgrade doesn't work. And that is configured in `connect` parameters.

As I've said, mayor use case for that is: pool read-only by default where all mutations use `retry` (i.e. where no mutation is possible without wrapping in `retry`), which I see as a pattern I would use in basically all my projects (except ad-hoc scripts). And this doesn't even need postgres replicas, so definitely possible through a single connection.

> Add an API to open readonly connections when the application knows it may connect to a replica: connect_readonly -> ReadonlyConnection. There will be no way to make a readonly connection a read/write one.

This is for quite limited set of use cases. The idea was that you can specify "I want replica with a lag of no more than 10 seconds". And edgedb keeps track of that lag and executes statement within either a replica that satisfies condition or within a master. Without reconnecting on the client. Hence replica usage and exact condition of that is a header for `Execute` packet.

**Reply by 1st1** [reply]
> The idea is that it's used like this:
>
> def hander(req, base_conn, *args):
>     user_conn = base_conn.with_session_config(username=req.session['username'])
>     handle_user_request(req, user_conn, *args)

Yes, but `base_conn` is can be used in parallel with handle user request. Just change your code to something like  

```python
async def hander(req, base_conn, *args):
    user_conn = await base_conn.with_session_config(username=req.session['username'])
    await asyncio.gather(
        handle_user_request(req, user_conn, *args),
        report_metrics(req, base_conn)
    )
```

I really don't like this because this would crash (or worse—get deadlocked) if there's one physical connection shared between `base_conn` and `user_conn`.


>> But we must guarantee that for all "virtual" connections created out of one connect() call we always have only one physical connection at a time.
>
> Why? The idea is that all of them can. If some can't we do that as a parameter to connect(). The "readonly" and "metadata" things are actually headers for Execute message, so set on per packet basis. So we could put it into query() signature if we want, it's just too weird to overload signature of that function too much.

Because it's easy to reason when `conn = connect(); user_code(conn)` means that only one active socket connection to EdgeDB will be used. Using multiple makes it impossible to reason about the code and the resources it might consume. Therefore I'm strong -1 on breaking this invariant.

> This is for quite limited set of use cases. The idea was that you can specify "I want replica with a lag of no more than 10 seconds". And edgedb keeps track of that lag and executes statement within either a replica that satisfies condition or within a master.

Sure, we can do that, I guess.  But having something like `connect_readonly(...) -> ReadonlyConnection` doesn't mean we can't implement your idea. It just makes sense from the typing standpoint, imo.

**Reply by tailhook** [reply]
> ```python
> async def hander(req, base_conn, *args):
>     user_conn = await base_conn.with_session_config(username=req.session['username'])
>     await asyncio.gather(
>         handle_user_request(req, user_conn, *args),
>         report_metrics(req, base_conn)
>     )
> ```
> 
> I really don't like this because this would crash (or worse—get deadlocked) if there's one physical connection shared between `base_conn` and `user_conn`.

Why would it crash? It would behave similarly to using single connection for parallel requests normally, just switch session data when needed.

Actually concurrent connection usage is the strong point of this RFC. Otherwise, you would need to carefully keep track of context, so that metrics are not submitted with user credentials.

Also you concentrate too much on single connection usage. Think of pool. Pool used this way would use optimal number of connections for that: i.e. probably would have single metrics connection around just for that purpose and prepare connection for this specific user. Or when there is a connection pressure, use the same connection for metrics as fast as user connection finished request.

> But we must guarantee that for all "virtual" connections created out of one connect() call we always have only one physical connection at a time.
>
> Because it's easy to reason when `conn = connect(); user_code(conn)` means that only one active socket connection to EdgeDB will be used. Using multiple makes it impossible to reason about the code and the resources it might consume. Therefore I'm strong -1 on breaking this invariant.
> 

If by "physical" connection you mean `app -> edgedb` connection. Then sure. Nothing in this spec contradicts that.

If by "physical" you mean `edgedb -> postgres`. Then no. We canceled that with "Ports RFC" (connection pooling).

Also: single connection usage is a subset of all usages. I'm more focused on Pool and single connection just replicates what is possible on the pool And yes for pool all of that is more useful that for single connection. But we don't want to have much more different APIs.

> Sure, we can do that, I guess. But having something like `connect_readonly(...) -> ReadonlyConnection` doesn't mean we can't implement your idea. It just makes sense from the typing standpoint, imo.


Sure `connect_readonly` will most likely be able to connect to readonly EdgeDB (which connects only to replicas). So yes, we will probably have that API.

Regarding `ReadonlyConnection` type, I'm not sure what is the primary purpose.

If it's just to mark that `def handle(conn: ReadonlyConnection)` doesn't do any read-write queries then:
1. We probably need `ReadonlyExecutor` that works on `Pool`, `Transaction`, `ReadonlyPool`, etc.
2. `ReadonlyExecutor` is probably a superclass of `Executor` so that you can pass normal connection to `handle()`
3. We may always return `ReadonlyConnection` and `ReadonlyPool` from `read_only()` method, but make them `read_only(allow_upgrade=True)` to allow uprading again (my use case of read-only in majority cases). But passing `handle(conn.read_only())` (`allow_upgrade=False` by default) would make sure that function doesn't mutate anything, and also doesn't preclude that other functions use original `conn` for mutations, and don't require extra physical connection.

Is this what you have in mind?

**Reply by 1st1** [reply]
>> I really don't like this because this would crash (or worse—get deadlocked) if there's one physical connection shared between base_conn and user_conn.
>
> Why would it crash? It would behave similarly to using single connection for parallel requests normally, just switch session data when needed.

We don't allow using single connection for parallel requests in any of our bindings. Our protocol is not designed for multiplexing multiple different requests on one "client <> edgedb" connection.

> Actually concurrent connection usage is the strong point of this RFC. Otherwise, you would need to carefully keep track of context, so that metrics are not submitted with user credentials.

Use connection pools for concurrent connection usage. A single connection shouldn't be used concurrently.

> Also you concentrate too much on single connection usage. Think of pool.

I've no problems with pools; but I'm against implementing "smart" strategies in pools. You acquire 10 connections and spawn 10 parallel tasks with each holding its own connection. Or every task would get a pool and acquire/release connection itself. Both use cases are totally OK and of conventional design.

> Or when there is a connection pressure, use the same connection for metrics as fast as user connection finished request.

But I'm against this ^^^. Or other "smart" connection use logic. IOW, I'm against the pool of connections, or a single connection, deciding internally how many "client <> edgedb" sockets they should use. That's hard for us to implement and hard for the users to reason about.

> If by "physical" connection you mean app -> edgedb connection. Then sure. Nothing in this spec contradicts that.

Yes, this is what I mean.

> If by "physical" you mean edgedb -> postgres. Then no. We canceled that with "Ports RFC" (connection pooling).

No, I didn't meat that, that's out of context of this discussion.


> Regarding ReadonlyConnection type, I'm not sure what is the primary purpose.

The primary purpose is to let users write code that explicitly accepts `ReadonlyConnection` and for compilers to verify that you either pass a connection created by `connect_readonly()` to that code, or connections created via `connection.read_only()`.

**Reply by jaclarke** [reply]
> We don't allow using single connection for parallel requests in any of our bindings. Our protocol is not designed for multiplexing multiple different requests on one "client <> edgedb" connection.

My understanding of the proposal here was that the `Connection` class acts as though it's a pool of size one, since it's defined using an internal `_SingletonPool` class near the start. If this is the case that it's just effectively a 'convenience class' for a pool of one physical connection, then no multiplexing would happen, and the parallel requests would just be executed sequentially right? This might be useful behaviour to stop users from accidentally multiplexing queries.

**Reply by 1st1** [reply]
> and the parallel requests would just be executed sequentially right? 

That's the thing - I don't want us to enable the "executed sequentially" part. That would require using queues internally and will slow everything down. Users are free to implement their own serialization of access to one connection the way they see fit, but it shouldn't be our business.

**Reply by 1st1** [reply]
Making a request with a connection that is processing another request currently results in an error. And I'd like to keep that with the new design.

**Reply by tailhook** [reply]
> We don't allow using single connection for parallel requests in any of our bindings. Our protocol is not designed for multiplexing multiple different requests on one "client <> edgedb" connection.
> 

Okay. First thing, currently this is bad:
```
>>> conn = l.run_until_complete(edgedb.async_connect(user='edgedb', database='edgedb'))
>>> print(l.run_until_complete(asyncio.gather(conn.query('SELECT 1+1'), conn.query('SELECT 2+2'))))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/work/edgedb/asyncio_con.py", line 102, in query
    return await self._protocol.execute_anonymous(
  File "edgedb/protocol/protocol.pyx", line 526, in execute_anonymous
    codecs = await self._parse(
  File "edgedb/protocol/protocol.pyx", line 256, in _parse
    self.fallthrough()
  File "edgedb/protocol/protocol.pyx", line 973, in edgedb.protocol.protocol.SansIOProtocol.fallthrough
    raise errors.ProtocolError(
edgedb.errors.ProtocolError: unexpected message type '1'
```
And we have to fix this regardless of how this proposal works out. This is because nothing stops users doing this even with current connection.

This is fixed by locking connection for the time of `query`. And locks are quite cheap for uncontended case.

We can probably leave `RawConnection` without locks if we measure that it suffers performance.

> > Actually concurrent connection usage is the strong point of this RFC. Otherwise, you would need to carefully keep track of context, so that metrics are not submitted with user credentials.
> 
> Use connection pools for concurrent connection usage. A single connection shouldn't be used concurrently.

1. This is not so obvious as one might think. For example there are database bindings that do request pipelining from multiple concurrent tasks ([for example aioredis](https://git.io/JTxnP)). Since python is single-threaded you can basically achieve highest performance by using a single connection for hundreds of users without connection pools.
2. That said since our protocol is (sometimes unneedlessly) more complex we can't do that. But our connection must not crash or have security issues when mistakenly used concurrently (see above).

> 
> > Also you concentrate too much on single connection usage. Think of pool.
> 
> I've no problems with pools; but I'm against implementing "smart" strategies in pools. You acquire 10 connections and spawn 10 parallel tasks with each holding its own connection. Or every task would get a pool and acquire/release connection itself. Both use cases are totally OK and of conventional design.
> 

Well, I don't think acquiring a connection should be a common case, since that risks holding connection longer than needed (i.e. when communicating to other external resources). So we should encourage either `pool.query()` or `pool.retry()` (latter obviously have some acquired connection).

That said, it might not be obvious, but I thought `pool.acquire()` should return (a proxy for) `RawConnection` not `Connection`. I mean, if you really know that you want that, it should be a single underlying socket without any reconnects possible.

> > Regarding ReadonlyConnection type, I'm not sure what is the primary purpose.
> 
> The primary purpose is to let users write code that explicitly accepts `ReadonlyConnection` and for compilers to verify that you either pass a connection created by `connect_readonly()` to that code, or connections created via `connection.read_only()`.

Yes this is different from what I've described. But why you need this?

I.e. if you have `def get_user_data(conn: ReadonlyConn)` which is a function that doesn't do writes. What is profit from calling it as `get_user_data(conn.read_only())` comparing to `get_user_data(conn)` ?

**Reply by 1st1** [reply]
> And we have to fix this regardless of how this proposal works out. This is because nothing stops users doing this even with current connection.

Oh shoot. This [is](https://github.com/edgedb/edgedb-js/blob/bf6cc5eeaecae5b2765ee9fcdc1359bd97e12a43/src/client.ts#L1013-L1020) [fixed](https://github.com/edgedb/edgedb-js/blob/bf6cc5eeaecae5b2765ee9fcdc1359bd97e12a43/src/client.ts#L990-L998) in our JS driver and I was 100% sure we do the same in edgedb-python. Thanks for pointing this out, I've created a bug to track this: https://github.com/edgedb/edgedb-python/issues/130.

>     That said since our protocol is (sometimes unneedlessly) more complex we can't do that. But our connection must not crash or have security issues when mistakenly used concurrently (see above).

Yeah.

> Well, I don't think acquiring a connection should be a common case, since that risks holding connection longer than needed (i.e. when communicating to other external resources). So we should encourage either pool.query() or pool.retry() (latter obviously have some acquired connection).

100% agree.

> That said, it might not be obvious, but I thought pool.acquire() should return (a proxy for) RawConnection not Connection. I mean, if you really know that you want that, it should be a single underlying socket without any reconnects possible.

Not sure about that. If I acquire a connection from the pool I want it to have the `con.retry()` method.

> I.e. if you have def get_user_data(conn: ReadonlyConn) which is a function that doesn't do writes. What is profit from calling it as get_user_data(conn.read_only()) comparing to get_user_data(conn) ?

The profit is in the vice-versa scenario. Suppose you have

`get_user_data(ReadonlyConnection)` and `set_user_data(ReadWriteConnection)`

then you wouldn't be able to call `set_user_data` from code where you only have a readonly connection.

Sorry, I should have clarified this earlier, but I myself was a little bit lost in details.

**Reply by tailhook** [reply]
> > That said, it might not be obvious, but I thought pool.acquire() should return (a proxy for) RawConnection not Connection. I mean, if you really know that you want that, it should be a single underlying socket without any reconnects possible.
> 
> Not sure about that. If I acquire a connection from the pool I want it to have the `con.retry()` method.
> 

I also not very sure about that, but there are a couple of questions:

1. Does `retry` steals a connection from pool or reconnects in-place? Former sounds better and will result in reconnect if all connections are busy anyway.
2. Do reconnect happen on `query()`. If this obj is similar to connection, then "yes"

But if we answer these questions as above:

1. It's unclear why ons acquired connection in the first place. Is it to reserve a "slot" in pool just for this task?
2. If it is "reserving slot" might be suboptimal because when disconnect happens it is essentially unreserved now.
3. And then this "reserved connection" still competes for database time on the EdgeDB level (as most installations are likely to allow more EdgeDB connections than postgres connections).

So the only time connection is really reserved is on a round of `try_transaction` inside a `retry` block.

So do you have any real use cases for `acquire/release` in mind? My understanding for now is that people think about `acquire/resolve` just because that was how databases traditionally worked generally and how transactions were implemented in particular.

> > I.e. if you have def get_user_data(conn: ReadonlyConn) which is a function that doesn't do writes. What is profit from calling it as get_user_data(conn.read_only()) comparing to get_user_data(conn) ?
> 
> The profit is in the vice-versa scenario. Suppose you have
> 
> `get_user_data(ReadonlyConnection)` and `set_user_data(ReadWriteConnection)`
> 
> then you wouldn't be able to call `set_user_data` from code where you only have a readonly connection.
> 

Good. So here is how I see this:

```
Connection.register(ReadonlyConnection)

def change_user(conn: Connection): ...
def get_user(conn: ReadonlyConnection): ...

def handler(conn: ReadonlyConnection):

    # This is okay for typechecker
    get_user(conn)

    # This is a typechecker error, but will work in runtime if Connection
    # is passed to handler
    change_user(conn)

conn = edgedb.connect()
ro_conn = edgedb.connect_readonly()
handler(conn)  # << okay as Connection is a subclass of ReadonlyConnection
handler(conn.read_only())  # << type check ok, but will crash in `change_user`
hanlder(ro_conn)  # << type check ok, but will crash in `change_user`
```

Now `ReadonlyConnection` does tree things:
1. Marks typechecker that readonly connection is okay
2. Propagates that info throught typechecker (i.e. only functions accepting read-only connections work)
3. Sends `READONLY` header that stops compiler from executing mutating queries

I think neither of these 3 goals conflicts with upgradeable connections:
```
conn = edgedb.connect()
ro_conn = edgedb.connect_readonly()
conn.with_mutations() # no-op
conn.read_only().with_mutations() # throw CantUpgrade
conn.read_only(allow_upgrade=True).with_mutations()  # okay returns Connection
ro_conn.with_mutations()  # throw CantUpgrade

# And disabling upgrades can't be overriden
ro_conn.read_only(allow_upgrade=True)  # throw CantUpgrade
conn.read_only().read_only(allow_upgrade=True)  # throw CantUpgrade
```

Does it solve your task as well as mine?

**Reply by 1st1** [reply]
@elprans please join the discussion. Please read the rfc and our discussion with Paul.

I'm in the middle of replying to Paul's latest comment right now. Please start another thread for your comments.

**Reply by elprans** [reply]
> It's unclear why ons acquired connection in the first place. Is it to reserve a "slot" in pool just for this task?

Yes.  The idea is that you're checking out a connection to process a request and you don't want a churn of re-configuring a new connection from a pool every time you want to make a query in a request handler.  Essentially:

```
with pool.acquire() as conn:
    conn.configure()
    ...
    conn.query()
    ...
    conn.query()
```

Remember, session configuration has a cost, you absolutely don't want to be emitting a bunch of `CONFIGURE SESSION` statements for each query in the same logical request handler.

> If it is "reserving slot" might be suboptimal because when disconnect happens it is essentially unreserved now.

Let's be clear: disconnects are an exception.  If one happened we reconnect or pick another connection from the pool.

> And then this "reserved connection" still competes for database time on the EdgeDB level 

Other than an open socket, there aren't very many resources you'd be using for a connection.  In any case it is more efficient to keep a socket open than it is to redo `CONFIGURE` calls.

**Reply by tailhook** [reply]
> > And then this "reserved connection" still competes for database time on the EdgeDB level
> 
> Other than an open socket, there aren't very many resources you'd be using for a connection. In any case it is more efficient to keep a socket open than it is to redo `CONFIGURE` calls.

It's not that clear, actually. For example:
```
with pool.acquire() as conn:
    conn.configure()
    conn.query()
    go_to_3rd_party_api()
    conn.query()
```
Here if `go_to_3rd_party_api()` could easily take 100ms or few seconds, even if on average it takes 1ms. It's definitely faster to reset session configuration two times (unless your database latency is slow). If app is doing 1000 requests per second (which is quite moderate performance), 10 such requests (i.e. less than 1 percent of traffic) makes a major bottleneck.

While having `with_config` semantics:
```
conn = pool.with_config(..)  # this is still pool, but who cares
conn.query()  # configuring connection, let's keep it around say for 10ms
go_to_3rd_party_api()  # connection is stolen by a different coroutine if this functions takes for too long
conn.query()  # use a cached connection, or configure connection again if above took long
```
The impact of configuring the second time is negligible if `go_to_3rd_party_api()` took long. And connection is cached if `got_to_3rd_party_api()` is quick. I.e. when its running time is comparable to delay of `configure()` request.

Note: in my opinion tightly related things in the query will mostly be either a single transaction or a single query (as our language is quite rich). So if user have split requests they probably wanted to insert some external (ans so potentially slow) calculation between queries.

**Reply by tailhook** [reply]
Will all that said, if we still return `Connection` not `RawConnection` from `acquire` we can keep both of the worlds:
```
with pool.acquire() as bare_conn:
    conn = bare_conn.with_config()
    ...
    conn.query()
    ...
    conn.query()
```
This will work exactly the same as your example (unless you're using `bare_conn` which you don't want to unless you know what you're doing). And I think `with_` is good enough to hint that you don't configure connection object in place but but rather use a return value. Other than that it exactly the same code complexity.

**Reply by 1st1** [reply]
> Does retry steals a connection from pool or reconnects in-place? Former sounds better and will result in reconnect if all connections are busy anyway.

If a connection was acquired from the pool it should reconnect in-place. In the future this would allow us to make the API more flexible, e.g. a user acquires a connection from the pool and adds additional layer of config to it: `con = (await pool.acquire()).with_config(...)`. That said we can be smart here -- if the acquired connection has no configuration on top of other connections in the pool we can take one of the other ones.

> It's unclear why ons acquired connection in the first place. Is it to reserve a "slot" in pool just for this task?

To add additional layers of session/config setup or to call `.read_only()` on it.  Also what @elprans said.

> Will all that said, if we still return Connection not RawConnection from acquire we can keep both of the worlds:

I'm not sure where the idea of returning `RawConnection` from `Pool.acquire()` is coming from. IMO if this is not needed and all `edgedb-python/js` should provide is a `connect_raw() -> RawConnection` method. Those who need custom pools/logic can build on top of that.

**Comment by 1st1** [comment]
IMO this is almost ready to become a proper RFC document. I think the only contention points are `connect_readonly()` (whether we need it or not) and few open questions about "raw connections". The latter is a non-issue -- let's add `connect_raw()` method to get them and that should be the only API for now.
